In [1]:
import pandas as pd
import numpy as np

# Feature engineering
import unicodedata
from nltk import ngrams, FreqDist
from unidecode import unidecode

## Data Cleaning

### Cleaning Indonesian Dataset

We have two Indonesian datasets, so let's compare them:

In [2]:
# Inspecting first Indonesian dataset
df_indo_1 = pd.read_excel('name_data/exigerData/EXGR_Indonesian names.xlsx')
df_indo_1.head()

Unnamed: 0            fullname  Family name  \
0   http://www.wikidata.org/entity/Q7313790    Bertrand Antolin          NaN   
1  http://www.wikidata.org/entity/Q15138877           Ray Rizal          NaN   
2  http://www.wikidata.org/entity/Q17411237  Samsuridjal Djauzi          NaN   
3  http://www.wikidata.org/entity/Q12497619          Max Arifin          NaN   
4   http://www.wikidata.org/entity/Q7475963              Donita          NaN   

   Given name  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [3]:
# Inspecting second Indonesian dataset
df_indo_2 = pd.read_excel('name_data/exigerData/EXGR_Indonesian names-2.xlsx')
df_indo_2.head()

Unnamed: 0      fullname Family name  \
0   http://www.wikidata.org/entity/Q7475963        Donita         NaN   
1   http://www.wikidata.org/entity/Q7645143     Supriyadi         NaN   
2   http://www.wikidata.org/entity/Q7844727  Triyaningsih         NaN   
3  http://www.wikidata.org/entity/Q12515781      Soerjadi         NaN   
4  http://www.wikidata.org/entity/Q12523244     Undunsyah         NaN   

     Given name Designation  \
0        Donita         NaN   
1     Supriyadi         NaN   
2  Triyaningsih         NaN   
3      Soerjadi         NaN   
4     Undunsyah         NaN   

                                             Comment  Unnamed: 6  Unnamed: 7  
0  It is common for Indonesians to have only 1 gi...         NaN           1  
1                                                NaN         NaN           1  
2                                                NaN         NaN           1  
3                                                NaN         NaN           1  
4                                                NaN         NaN           1

Are the names in the datasets the same?

In [4]:
try:
    pd.testing.assert_series_equal(df_indo_1['fullname'], df_indo_2['fullname'], check_like = True)
except:
    print("Names are not the same")

Names are not the same


We will investigate this further later. For now, let's look at the comments in `df_indo_2`:

In [5]:
print(df_indo_2['Comment'].unique()[:5])
print('\nNumber of names with comments:', len(df_indo_2['Comment'].unique()))
df_indo_2.shape

['It is common for Indonesians to have only 1 given name without family name.'
 nan 'Real name: Mike Lucock'
 'This is a Javanese ancient name. The name is not used anymore. This person lived in 1028 - 1035.'
 'Royal family name of Java identified by the designation name.']

Number of names with comments: 34


(21727, 8)

Since these names seem uncommon exceptions and there are only a few of them, we will remove them to make the data cleaner:

In [6]:
df_indo_2 = df_indo_2[pd.isnull(df_indo_2['Comment'])]
df_indo_2.reset_index()
print(df_indo_2.shape)
df_indo_2['Comment'].unique()

(21689, 8)


array([nan], dtype=object)

Making dataframes consist only of the `fullname` column:

In [7]:
df_indo_1 = pd.DataFrame(df_indo_1['fullname'])
df_indo_2 = pd.DataFrame(df_indo_2['fullname'])
print('df_indo_1 shape:', df_indo_1.shape)
print('df_indo_2 shape:', df_indo_2.shape)

df_indo_1 shape: (21730, 1)
df_indo_2 shape: (21689, 1)


Combining both dataframes into `df_indo`:

In [8]:
# Combining datasets
df_indo = pd.concat([df_indo_1, df_indo_2], ignore_index = True)
print('Shape of combined data:', df_indo.shape)
df_indo.head()

Shape of combined data: (43419, 1)


fullname
0    Bertrand Antolin
1           Ray Rizal
2  Samsuridjal Djauzi
3          Max Arifin
4              Donita

Looking for duplicate entries:

In [9]:
print(len(df_indo[df_indo.duplicated()]))

32085


Since around 3/4 of the combined data consists of duplicate entries, the datasets are very similar. Let's just use `df_indo_2` since we removed the names with comments from it:

In [10]:
df_indo = df_indo_2
print(df_indo.shape)
df_indo.head()

(21689, 1)


fullname
1     Supriyadi
2  Triyaningsih
3      Soerjadi
4     Undunsyah
5      Soeripto

Are there duplicates?

In [11]:
len(df_indo[df_indo.duplicated()])

10376

Let's drop them:

In [12]:
# Dropping duplicates
df_indo.drop_duplicates(inplace = True, ignore_index = True)
print(np.any(df_indo.duplicated()))
df_indo.shape

False


(11313, 1)

Let's see if there are any null entries for `fullname`:

In [13]:
np.any(df_indo['fullname'].isnull())

False

What about entries that aren't alphanumeric aside from spaces?

In [14]:
non_alnum_names_indo = [name for name in df_indo['fullname'] if not str.isalnum(name.replace(' ', ''))]
print(len(non_alnum_names_indo))
non_alnum_names_indo[-10:]

590


['I. B. Rai Dharmawijaya Mantra',
 'Mansoer Daoed Dt. Palimo Kayo',
 'Habib Anis bin Alwi al-Habsyi',
 'Abu Bakar (bupati Bandung Barat)',
 'Muhammad Arsyad (pemain sepak bola)',
 'Muhammad Hamzah (personil The Fly)',
 'A. A. Ngurah Oka Ratmadi',
 'Firmansyah (pemain sepak bola kelahiran 1995)',
 'Mayjen TNI (Purn) Thomas Albert Umboh',
 'Prof. Dr. Ir. Antonius Suwanto, M.Sc']

Some names have parentheses. Let's see exactly how many:

In [15]:
paren_names_indo = [name for name in df_indo['fullname'] if name.find('(') != -1]
print(len(paren_names_indo))
paren_names_indo[:5]

67


['Basuki (pelawak)',
 'Mulyadi (politisi)',
 'Fatahillah (politisi)',
 'Trinity (penulis)',
 'Mentari (aktris)']

Removing the parentheses from the names and seeing how many names are non-alphanumeric:

In [16]:
df_indo['fullname'] = df_indo['fullname'].apply(lambda name: name if name.find('(') == -1 else name[:name.find('(') - 1])
non_alnum_names_indo = [name for name in df_indo['fullname'] if not str.isalnum(name.replace(' ', ''))]
print(len(non_alnum_names_indo))

527


The number of names left isn't exactly 590 - 67 = 523 because some names with parentheses had other punctuation in it. Let's double check that there are no names with parentheses in the `fullname` column:

In [17]:
print([name for name in df_indo['fullname'] if name.find('(') != -1])
print([name for name in df_indo['fullname'] if name.find(')') != -1])

[]
[]


Let's make every name lowercase to help us with feature engineering. We will first save the original `fullname` column into a new column, `original_fullname`, and then turn `fullname` into lowercase:

In [18]:
df_indo['original_fullname'] = df_indo['fullname']
df_indo['fullname'] = df_indo['fullname'].apply(str.lower)
df_indo

fullname  \
0                                              supriyadi   
1                                           triyaningsih   
2                                               soerjadi   
3                                              undunsyah   
4                                               soeripto   
...                                                  ...   
11308                                         mayjen tni   
11309          sultan amaluddin al sani perkasa alamsyah   
11310               prof. dr. ir. antonius suwanto, m.sc   
11311  kanjeng raden tumengung mas ariya purnama hadi...   
11312  lord kanjeng baginda atep bin lord dari segala...   

                                       original_fullname  
0                                              Supriyadi  
1                                           Triyaningsih  
2                                               Soerjadi  
3                                              Undunsyah  
4                                               Soeripto  
...                                                  ...  
11308                                         Mayjen TNI  
11309          Sultan Amaluddin Al Sani Perkasa Alamsyah  
11310               Prof. Dr. Ir. Antonius Suwanto, M.Sc  
11311  Kanjeng Raden Tumengung Mas Ariya Purnama Hadi...  
11312  Lord Kanjeng Baginda Atep Bin Lord Dari Segala...  

[11313 rows x 2 columns]

It's possible that we have some duplicate names after removing parentheses and turning names lowercase, so let's check for that:

In [19]:
print(len(df_indo[df_indo.duplicated()]))
print(df_indo.shape)
df_indo[df_indo.duplicated()]

63
(11313, 2)


fullname original_fullname
1676            basuki            Basuki
1810           mulyadi           Mulyadi
2716        fatahillah        Fatahillah
3080           trinity           Trinity
4798           mentari           Mentari
...                ...               ...
11285        abu bakar         Abu Bakar
11286  muhammad arsyad   Muhammad Arsyad
11288  muhammad hamzah   Muhammad Hamzah
11300       firmansyah        Firmansyah
11308       mayjen tni        Mayjen TNI

[63 rows x 2 columns]

Removing leading and trailing whitespace from names just in case:

In [20]:
df_indo['fullname'] = df_indo['fullname'].apply(str.strip)

Removing duplicates again:

In [21]:
df_indo.drop_duplicates(inplace = True, ignore_index = True)
print(np.any(df_indo.duplicated()))
df_indo.shape

False


(11250, 2)

### Cleaning Malay Dataset

Let's look at the Malay dataset now:

In [22]:
df_malay = pd.read_excel('name_data/exigerData/EXGR_Malay names.xlsx')
df_malay.head()

Unnamed: 0                   fullname  \
0   http://www.wikidata.org/entity/Q4769705              Annuar Rapaee   
1  http://www.wikidata.org/entity/Q31186972                  Tash Yong   
2    http://www.wikidata.org/entity/Q468519                  Fatmawati   
3   http://www.wikidata.org/entity/Q4736793                 Alto Linus   
4  http://www.wikidata.org/entity/Q28837179  Mohamad Izzat Abdul Halil   

   Family name  Given name  
0          NaN         NaN  
1          NaN         NaN  
2          NaN         NaN  
3          NaN         NaN  
4          NaN         NaN

Restricting the dataframe to just the `fullname` column:

In [23]:
df_malay = pd.DataFrame(df_malay['fullname'])
df_malay.head()

fullname
0              Annuar Rapaee
1                  Tash Yong
2                  Fatmawati
3                 Alto Linus
4  Mohamad Izzat Abdul Halil

Handling duplicate values:

In [24]:
# Checking for duplicate entries
print(len(df_malay[df_malay.duplicated()]))
df_malay.shape

2016


(4930, 1)

In [25]:
# Removing duplicate entries
df_malay.drop_duplicates(inplace = True, ignore_index = True)
print(np.any(df_malay.duplicated()))
df_malay.shape

False


(2914, 1)

Are there any null entries in `fullname`?

In [26]:
print(np.any(df_malay['fullname'].isnull()))

False


Are any names non-alphanumeric aside from spaces?

In [27]:
non_alnum_names_malay = [name for name in df_malay['fullname'] if not str.isalnum(name.replace(' ', ''))]
print(len(non_alnum_names_malay))
non_alnum_names_malay[-10:]

200


['Adenan Hj. Satem',
 'P. Balasubramaniam',
 "Abdullah Ma'ayat Shah",
 "Mohd Shahril Sa'ari",
 'E. E. C. Thuraisingham',
 'Ibrahim Ali (Malaysia)',
 'Faradina Mohd. Nadzir',
 'Aril (AF 7)',
 'Sultan Hisamuddin Alam Shah Al-Haj ibni Almarhum Sultan Alaeddin Sulaiman Shah',
 "Ismail Faruqi Asha'ri"]

Again, we have parentheses at the end, so let's remove them:

In [28]:
df_malay['fullname'] = df_malay['fullname'].apply(lambda name: name if name.find('(') == -1 else name[:name.find('(') - 1])
non_alnum_names_malay = [name for name in df_malay['fullname'] if not str.isalnum(name.replace(' ', ''))]
print(len(non_alnum_names_malay))

192


Turning names lowercase:

In [29]:
df_malay['original_fullname'] = df_malay['fullname']
df_malay['fullname'] = df_malay['fullname'].apply(str.lower)
df_malay

fullname  \
0                            annuar rapaee   
1                                tash yong   
2                                fatmawati   
3                               alto linus   
4                mohamad izzat abdul halil   
...                                    ...   
2909                   salmah binti ismail   
2910              abdul aziz sheikh fadzir   
2911  sultan ibrahim ibni sultan abu bakar   
2912                          ahmad yaakob   
2913                        jailani johari   

                         original_fullname  
0                            Annuar Rapaee  
1                                Tash Yong  
2                                Fatmawati  
3                               Alto Linus  
4                Mohamad Izzat Abdul Halil  
...                                    ...  
2909                   Salmah binti Ismail  
2910              Abdul Aziz Sheikh Fadzir  
2911  Sultan Ibrahim ibni Sultan Abu Bakar  
2912                          Ahmad Yaakob  
2913                        Jailani Johari  

[2914 rows x 2 columns]

Removing leading and trailing whitespace from names just in case:

In [30]:
df_malay['fullname'] = df_malay['fullname'].apply(str.strip)

Handling duplicates:

In [31]:
# Finding number of duplicates
print(len(df_malay[df_malay.duplicated()]))
df_malay.shape

6


(2914, 2)

In [32]:
# Removing duplicates
df_malay.drop_duplicates(inplace = True, ignore_index = True)
print(np.any(df_malay.duplicated()))
df_malay.shape

False


(2908, 2)

## Feature Engineering

### Indonesian Dataset Feature Engineering

Here are the features we want to include:
1. `alphabet` - which alphabet is being used - unicode library
2. `char_ngrams` - character n-grams (unigram, bigram, trigram, interpolated) - `nltk.utils.ngrams()`
3. `word_ngrams` - word n-grams
4. `a_hat_freq` - binary feature: does it exist in the given name? - if too much work to have EACH accent character as a feature, can combine
5. `name_length` - length of name (total number of letters in full name)
6. `avg_token_length` - average length of a token (the avg number of characters in each word of the name)
7. `num_tokens` - how many “words” are in each name
8. `transliteration` - transliteration
9. `period_freq`, `dash_freq`, `apostrophe_freq` - frequency of punctuation
10. `space_freq` - how many spaces are in each name

Let's start with the Indonesian dataset.

In [33]:
# 1. alphabet
def get_alphabet(name):
    name_as_alphabet_list = []
    for char in name:
        name_as_alphabet_list.append(unicodedata.name(char).split(' ')[0])
    return name_as_alphabet_list

df_indo['alphabet'] = df_indo['fullname'].apply(get_alphabet)
df_indo.head()

fullname original_fullname  \
0     supriyadi         Supriyadi   
1  triyaningsih      Triyaningsih   
2      soerjadi          Soerjadi   
3     undunsyah         Undunsyah   
4      soeripto          Soeripto   

                                            alphabet  
0  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...  
1  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...  
2  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...  
3  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...  
4  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...

In [34]:
# 2. unigrams, bigrams, trigrams, char_ngrams
df_indo['unigrams'] = df_indo['fullname'].apply(lambda name: list(name))
df_indo['bigrams'] = df_indo['fullname'].apply(lambda name: list(ngrams(list(name), 2)))
df_indo['trigrams'] = df_indo['fullname'].apply(lambda name: list(ngrams(list(name), 3)))
df_indo['char_ngrams'] = df_indo['unigrams'] + df_indo['bigrams'] + df_indo['trigrams']
df_indo.head()

fullname original_fullname  \
0     supriyadi         Supriyadi   
1  triyaningsih      Triyaningsih   
2      soerjadi          Soerjadi   
3     undunsyah         Undunsyah   
4      soeripto          Soeripto   

                                            alphabet  \
0  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
1  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
2  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
3  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
4  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   

                               unigrams  \
0           [s, u, p, r, i, y, a, d, i]   
1  [t, r, i, y, a, n, i, n, g, s, i, h]   
2              [s, o, e, r, j, a, d, i]   
3           [u, n, d, u, n, s, y, a, h]   
4              [s, o, e, r, i, p, t, o]   

                                             bigrams  \
0  [(s, u), (u, p), (p, r), (r, i), (i, y), (y, a...   
1  [(t, r), (r, i), (i, y), (y, a), (a, n), (n, i...   
2  [(s, o), (o, e), (e, r), (r, j), (j, a), (a, d...   
3  [(u, n), (n, d), (d, u), (u, n), (n, s), (s, y...   
4  [(s, o), (o, e), (e, r), (r, i), (i, p), (p, t...   

                                            trigrams  \
0  [(s, u, p), (u, p, r), (p, r, i), (r, i, y), (...   
1  [(t, r, i), (r, i, y), (i, y, a), (y, a, n), (...   
2  [(s, o, e), (o, e, r), (e, r, j), (r, j, a), (...   
3  [(u, n, d), (n, d, u), (d, u, n), (u, n, s), (...   
4  [(s, o, e), (o, e, r), (e, r, i), (r, i, p), (...   

                                         char_ngrams  
0  [s, u, p, r, i, y, a, d, i, (s, u), (u, p), (p...  
1  [t, r, i, y, a, n, i, n, g, s, i, h, (t, r), (...  
2  [s, o, e, r, j, a, d, i, (s, o), (o, e), (e, r...  
3  [u, n, d, u, n, s, y, a, h, (u, n), (n, d), (d...  
4  [s, o, e, r, i, p, t, o, (s, o), (o, e), (e, r...

In [35]:
# 3. word_ngrams
df_indo['word_ngrams'] = df_indo['fullname'].apply(lambda name: name.split())
df_indo.tail()

fullname  \
11245                 edwel yusri datuak rajo gampo alam   
11246          sultan amaluddin al sani perkasa alamsyah   
11247               prof. dr. ir. antonius suwanto, m.sc   
11248  kanjeng raden tumengung mas ariya purnama hadi...   
11249  lord kanjeng baginda atep bin lord dari segala...   

                                       original_fullname  \
11245                 Edwel Yusri Datuak Rajo Gampo Alam   
11246          Sultan Amaluddin Al Sani Perkasa Alamsyah   
11247               Prof. Dr. Ir. Antonius Suwanto, M.Sc   
11248  Kanjeng Raden Tumengung Mas Ariya Purnama Hadi...   
11249  Lord Kanjeng Baginda Atep Bin Lord Dari Segala...   

                                                alphabet  \
11245  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...   
11246  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...   
11247  [LATIN, LATIN, LATIN, LATIN, FULL, SPACE, LATI...   
11248  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
11249  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...   

                                                unigrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                                 bigrams  \
11245  [(e, d), (d, w), (w, e), (e, l), (l,  ), ( , y...   
11246  [(s, u), (u, l), (l, t), (t, a), (a, n), (n,  ...   
11247  [(p, r), (r, o), (o, f), (f, .), (.,  ), ( , d...   
11248  [(k, a), (a, n), (n, j), (j, e), (e, n), (n, g...   
11249  [(l, o), (o, r), (r, d), (d,  ), ( , k), (k, a...   

                                                trigrams  \
11245  [(e, d, w), (d, w, e), (w, e, l), (e, l,  ), (...   
11246  [(s, u, l), (u, l, t), (l, t, a), (t, a, n), (...   
11247  [(p, r, o), (r, o, f), (o, f, .), (f, .,  ), (...   
11248  [(k, a, n), (a, n, j), (n, j, e), (j, e, n), (...   
11249  [(l, o, r), (o, r, d), (r, d,  ), (d,  , k), (...   

                                             char_ngrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                             word_ngrams  
11245          [edwel, yusri, datuak, rajo, gampo, alam]  
11246   [sultan, amaluddin, al, sani, perkasa, alamsyah]  
11247        [prof., dr., ir., antonius, suwanto,, m.sc]  
11248  [kanjeng, raden, tumengung, mas, ariya, purnam...  
11249  [lord, kanjeng, baginda, atep, bin, lord, dari...

We will skip #4, `a_hat_freq`, for now because it is difficult to account for across multiple languages.

In [36]:
# 5. name_length
df_indo['name_length'] = df_indo['fullname'].apply(len)
df_indo.head()

fullname original_fullname  \
0     supriyadi         Supriyadi   
1  triyaningsih      Triyaningsih   
2      soerjadi          Soerjadi   
3     undunsyah         Undunsyah   
4      soeripto          Soeripto   

                                            alphabet  \
0  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
1  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
2  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
3  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
4  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   

                               unigrams  \
0           [s, u, p, r, i, y, a, d, i]   
1  [t, r, i, y, a, n, i, n, g, s, i, h]   
2              [s, o, e, r, j, a, d, i]   
3           [u, n, d, u, n, s, y, a, h]   
4              [s, o, e, r, i, p, t, o]   

                                             bigrams  \
0  [(s, u), (u, p), (p, r), (r, i), (i, y), (y, a...   
1  [(t, r), (r, i), (i, y), (y, a), (a, n), (n, i...   
2  [(s, o), (o, e), (e, r), (r, j), (j, a), (a, d...   
3  [(u, n), (n, d), (d, u), (u, n), (n, s), (s, y...   
4  [(s, o), (o, e), (e, r), (r, i), (i, p), (p, t...   

                                            trigrams  \
0  [(s, u, p), (u, p, r), (p, r, i), (r, i, y), (...   
1  [(t, r, i), (r, i, y), (i, y, a), (y, a, n), (...   
2  [(s, o, e), (o, e, r), (e, r, j), (r, j, a), (...   
3  [(u, n, d), (n, d, u), (d, u, n), (u, n, s), (...   
4  [(s, o, e), (o, e, r), (e, r, i), (r, i, p), (...   

                                         char_ngrams     word_ngrams  \
0  [s, u, p, r, i, y, a, d, i, (s, u), (u, p), (p...     [supriyadi]   
1  [t, r, i, y, a, n, i, n, g, s, i, h, (t, r), (...  [triyaningsih]   
2  [s, o, e, r, j, a, d, i, (s, o), (o, e), (e, r...      [soerjadi]   
3  [u, n, d, u, n, s, y, a, h, (u, n), (n, d), (d...     [undunsyah]   
4  [s, o, e, r, i, p, t, o, (s, o), (o, e), (e, r...      [soeripto]   

   name_length  
0            9  
1           12  
2            8  
3            9  
4            8

In [37]:
# 6. avg_token_length
tokens = df_indo['fullname'].apply(lambda name: name.split(' '))
print(tokens[-5:], '\n')
token_lengths = tokens.apply(lambda token_list: [len(token) for token in token_list])
print(token_lengths[-5:])
df_indo['avg_token_length'] = token_lengths.apply(np.mean)
df_indo.tail()

11245            [edwel, yusri, datuak, rajo, gampo, alam]
11246     [sultan, amaluddin, al, sani, perkasa, alamsyah]
11247          [prof., dr., ir., antonius, suwanto,, m.sc]
11248    [kanjeng, raden, tumengung, mas, ariya, purnam...
11249    [lord, kanjeng, baginda, atep, bin, lord, dari...
Name: fullname, dtype: object 

11245                      [5, 5, 6, 4, 5, 4]
11246                      [6, 9, 2, 4, 7, 8]
11247                      [5, 3, 3, 8, 8, 4]
11248                  [7, 5, 9, 3, 5, 7, 11]
11249    [4, 7, 7, 4, 3, 4, 4, 6, 4, 7, 7, 8]
Name: fullname, dtype: object


fullname  \
11245                 edwel yusri datuak rajo gampo alam   
11246          sultan amaluddin al sani perkasa alamsyah   
11247               prof. dr. ir. antonius suwanto, m.sc   
11248  kanjeng raden tumengung mas ariya purnama hadi...   
11249  lord kanjeng baginda atep bin lord dari segala...   

                                       original_fullname  \
11245                 Edwel Yusri Datuak Rajo Gampo Alam   
11246          Sultan Amaluddin Al Sani Perkasa Alamsyah   
11247               Prof. Dr. Ir. Antonius Suwanto, M.Sc   
11248  Kanjeng Raden Tumengung Mas Ariya Purnama Hadi...   
11249  Lord Kanjeng Baginda Atep Bin Lord Dari Segala...   

                                                alphabet  \
11245  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...   
11246  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...   
11247  [LATIN, LATIN, LATIN, LATIN, FULL, SPACE, LATI...   
11248  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
11249  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...   

                                                unigrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                                 bigrams  \
11245  [(e, d), (d, w), (w, e), (e, l), (l,  ), ( , y...   
11246  [(s, u), (u, l), (l, t), (t, a), (a, n), (n,  ...   
11247  [(p, r), (r, o), (o, f), (f, .), (.,  ), ( , d...   
11248  [(k, a), (a, n), (n, j), (j, e), (e, n), (n, g...   
11249  [(l, o), (o, r), (r, d), (d,  ), ( , k), (k, a...   

                                                trigrams  \
11245  [(e, d, w), (d, w, e), (w, e, l), (e, l,  ), (...   
11246  [(s, u, l), (u, l, t), (l, t, a), (t, a, n), (...   
11247  [(p, r, o), (r, o, f), (o, f, .), (f, .,  ), (...   
11248  [(k, a, n), (a, n, j), (n, j, e), (j, e, n), (...   
11249  [(l, o, r), (o, r, d), (r, d,  ), (d,  , k), (...   

                                             char_ngrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                             word_ngrams  name_length  \
11245          [edwel, yusri, datuak, rajo, gampo, alam]           34   
11246   [sultan, amaluddin, al, sani, perkasa, alamsyah]           41   
11247        [prof., dr., ir., antonius, suwanto,, m.sc]           36   
11248  [kanjeng, raden, tumengung, mas, ariya, purnam...           53   
11249  [lord, kanjeng, baginda, atep, bin, lord, dari...           76   

       avg_token_length  
11245          4.833333  
11246          6.000000  
11247          5.166667  
11248          6.714286  
11249          5.416667

Finding `num_tokens`:

In [38]:
# 7. num_tokens
df_indo['num_tokens'] = df_indo['fullname'].apply(lambda name: len(name.split(' ')))
df_indo.tail()

fullname  \
11245                 edwel yusri datuak rajo gampo alam   
11246          sultan amaluddin al sani perkasa alamsyah   
11247               prof. dr. ir. antonius suwanto, m.sc   
11248  kanjeng raden tumengung mas ariya purnama hadi...   
11249  lord kanjeng baginda atep bin lord dari segala...   

                                       original_fullname  \
11245                 Edwel Yusri Datuak Rajo Gampo Alam   
11246          Sultan Amaluddin Al Sani Perkasa Alamsyah   
11247               Prof. Dr. Ir. Antonius Suwanto, M.Sc   
11248  Kanjeng Raden Tumengung Mas Ariya Purnama Hadi...   
11249  Lord Kanjeng Baginda Atep Bin Lord Dari Segala...   

                                                alphabet  \
11245  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...   
11246  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...   
11247  [LATIN, LATIN, LATIN, LATIN, FULL, SPACE, LATI...   
11248  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
11249  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...   

                                                unigrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                                 bigrams  \
11245  [(e, d), (d, w), (w, e), (e, l), (l,  ), ( , y...   
11246  [(s, u), (u, l), (l, t), (t, a), (a, n), (n,  ...   
11247  [(p, r), (r, o), (o, f), (f, .), (.,  ), ( , d...   
11248  [(k, a), (a, n), (n, j), (j, e), (e, n), (n, g...   
11249  [(l, o), (o, r), (r, d), (d,  ), ( , k), (k, a...   

                                                trigrams  \
11245  [(e, d, w), (d, w, e), (w, e, l), (e, l,  ), (...   
11246  [(s, u, l), (u, l, t), (l, t, a), (t, a, n), (...   
11247  [(p, r, o), (r, o, f), (o, f, .), (f, .,  ), (...   
11248  [(k, a, n), (a, n, j), (n, j, e), (j, e, n), (...   
11249  [(l, o, r), (o, r, d), (r, d,  ), (d,  , k), (...   

                                             char_ngrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                             word_ngrams  name_length  \
11245          [edwel, yusri, datuak, rajo, gampo, alam]           34   
11246   [sultan, amaluddin, al, sani, perkasa, alamsyah]           41   
11247        [prof., dr., ir., antonius, suwanto,, m.sc]           36   
11248  [kanjeng, raden, tumengung, mas, ariya, purnam...           53   
11249  [lord, kanjeng, baginda, atep, bin, lord, dari...           76   

       avg_token_length  num_tokens  
11245          4.833333           6  
11246          6.000000           6  
11247          5.166667           6  
11248          6.714286           7  
11249          5.416667          12

In [39]:
# 8. transliteration
df_indo['transliteration'] = df_indo['fullname'].apply(lambda name: unidecode(name))
df_indo[df_indo['fullname'] != df_indo['transliteration']][['fullname', 'transliteration']]

fullname  \
1010                         cristian gonzáles   
1140                              śri ajñadewi   
1211                              ipe ma’aroef   
2868                              fauzie as’ad   
3181                                kiki maría   
3263                          abdullah wasi’an   
3302                             empu prapañca   
4598                        manuel carrascalão   
6799                               jaëll hattu   
6808                       berthold damshäuser   
8037                            anneke grönloh   
8404                            joesoef sou’yb   
8556                              fauzie as’ad   
8876                 guilherme maria gonçalves   
10203                 mário viegas carrascalão   
10550                   sarah tüffahati mornov   
10895                    joão da costa tavares   
10900                  arnaldo dos reis araújo   
11190  andré joseph guillaume henri kostermans   
11244   sri jayendra dyah dewi prajña paramita   

                               transliteration  
1010                         cristian gonzales  
1140                              sri ajnadewi  
1211                              ipe ma'aroef  
2868                              fauzie as'ad  
3181                                kiki maria  
3263                          abdullah wasi'an  
3302                             empu prapanca  
4598                        manuel carrascalao  
6799                               jaell hattu  
6808                       berthold damshauser  
8037                            anneke gronloh  
8404                            joesoef sou'yb  
8556                              fauzie as'ad  
8876                 guilherme maria goncalves  
10203                 mario viegas carrascalao  
10550                   sarah tuffahati mornov  
10895                    joao da costa tavares  
10900                  arnaldo dos reis araujo  
11190  andre joseph guillaume henri kostermans  
11244   sri jayendra dyah dewi prajna paramita

In [40]:
# 9. period_freq, dash_freq, apostrophe_freq
df_indo['period_freq'] = df_indo['fullname'].apply(lambda name: name.count('.'))
df_indo['dash_freq'] = df_indo['fullname'].apply(lambda name: name.count('-'))
df_indo['apostrophe_freq'] = df_indo['fullname'].apply(lambda name: name.count('\''))
df_indo.tail()

fullname  \
11245                 edwel yusri datuak rajo gampo alam   
11246          sultan amaluddin al sani perkasa alamsyah   
11247               prof. dr. ir. antonius suwanto, m.sc   
11248  kanjeng raden tumengung mas ariya purnama hadi...   
11249  lord kanjeng baginda atep bin lord dari segala...   

                                       original_fullname  \
11245                 Edwel Yusri Datuak Rajo Gampo Alam   
11246          Sultan Amaluddin Al Sani Perkasa Alamsyah   
11247               Prof. Dr. Ir. Antonius Suwanto, M.Sc   
11248  Kanjeng Raden Tumengung Mas Ariya Purnama Hadi...   
11249  Lord Kanjeng Baginda Atep Bin Lord Dari Segala...   

                                                alphabet  \
11245  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...   
11246  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...   
11247  [LATIN, LATIN, LATIN, LATIN, FULL, SPACE, LATI...   
11248  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
11249  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...   

                                                unigrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                                 bigrams  \
11245  [(e, d), (d, w), (w, e), (e, l), (l,  ), ( , y...   
11246  [(s, u), (u, l), (l, t), (t, a), (a, n), (n,  ...   
11247  [(p, r), (r, o), (o, f), (f, .), (.,  ), ( , d...   
11248  [(k, a), (a, n), (n, j), (j, e), (e, n), (n, g...   
11249  [(l, o), (o, r), (r, d), (d,  ), ( , k), (k, a...   

                                                trigrams  \
11245  [(e, d, w), (d, w, e), (w, e, l), (e, l,  ), (...   
11246  [(s, u, l), (u, l, t), (l, t, a), (t, a, n), (...   
11247  [(p, r, o), (r, o, f), (o, f, .), (f, .,  ), (...   
11248  [(k, a, n), (a, n, j), (n, j, e), (j, e, n), (...   
11249  [(l, o, r), (o, r, d), (r, d,  ), (d,  , k), (...   

                                             char_ngrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                             word_ngrams  name_length  \
11245          [edwel, yusri, datuak, rajo, gampo, alam]           34   
11246   [sultan, amaluddin, al, sani, perkasa, alamsyah]           41   
11247        [prof., dr., ir., antonius, suwanto,, m.sc]           36   
11248  [kanjeng, raden, tumengung, mas, ariya, purnam...           53   
11249  [lord, kanjeng, baginda, atep, bin, lord, dari...           76   

       avg_token_length  num_tokens  \
11245          4.833333           6   
11246          6.000000           6   
11247          5.166667           6   
11248          6.714286           7   
11249          5.416667          12   

                                         transliteration  period_freq  \
11245                 edwel yusri datuak rajo gampo alam            0   
11246          sultan amaluddin al sani perkasa alamsyah            0   
11247               prof. dr. ir. antonius suwanto, m.sc            4   
11248  kanjeng raden tumengung mas ariya purnama hadi...            0   
11249  lord kanjeng baginda atep bin lord dari segala...            0   

       dash_freq  apostrophe_freq  
11245          0                0  
11246          0                0  
11247          0                0  
11248          0                0  
11249          0                0

In [41]:
# 10. space_freq
df_indo['space_freq'] = df_indo['fullname'].apply(lambda name: name.count(' '))
df_indo.tail()

fullname  \
11245                 edwel yusri datuak rajo gampo alam   
11246          sultan amaluddin al sani perkasa alamsyah   
11247               prof. dr. ir. antonius suwanto, m.sc   
11248  kanjeng raden tumengung mas ariya purnama hadi...   
11249  lord kanjeng baginda atep bin lord dari segala...   

                                       original_fullname  \
11245                 Edwel Yusri Datuak Rajo Gampo Alam   
11246          Sultan Amaluddin Al Sani Perkasa Alamsyah   
11247               Prof. Dr. Ir. Antonius Suwanto, M.Sc   
11248  Kanjeng Raden Tumengung Mas Ariya Purnama Hadi...   
11249  Lord Kanjeng Baginda Atep Bin Lord Dari Segala...   

                                                alphabet  \
11245  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...   
11246  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...   
11247  [LATIN, LATIN, LATIN, LATIN, FULL, SPACE, LATI...   
11248  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
11249  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...   

                                                unigrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                                 bigrams  \
11245  [(e, d), (d, w), (w, e), (e, l), (l,  ), ( , y...   
11246  [(s, u), (u, l), (l, t), (t, a), (a, n), (n,  ...   
11247  [(p, r), (r, o), (o, f), (f, .), (.,  ), ( , d...   
11248  [(k, a), (a, n), (n, j), (j, e), (e, n), (n, g...   
11249  [(l, o), (o, r), (r, d), (d,  ), ( , k), (k, a...   

                                                trigrams  \
11245  [(e, d, w), (d, w, e), (w, e, l), (e, l,  ), (...   
11246  [(s, u, l), (u, l, t), (l, t, a), (t, a, n), (...   
11247  [(p, r, o), (r, o, f), (o, f, .), (f, .,  ), (...   
11248  [(k, a, n), (a, n, j), (n, j, e), (j, e, n), (...   
11249  [(l, o, r), (o, r, d), (r, d,  ), (d,  , k), (...   

                                             char_ngrams  \
11245  [e, d, w, e, l,  , y, u, s, r, i,  , d, a, t, ...   
11246  [s, u, l, t, a, n,  , a, m, a, l, u, d, d, i, ...   
11247  [p, r, o, f, .,  , d, r, .,  , i, r, .,  , a, ...   
11248  [k, a, n, j, e, n, g,  , r, a, d, e, n,  , t, ...   
11249  [l, o, r, d,  , k, a, n, j, e, n, g,  , b, a, ...   

                                             word_ngrams  name_length  \
11245          [edwel, yusri, datuak, rajo, gampo, alam]           34   
11246   [sultan, amaluddin, al, sani, perkasa, alamsyah]           41   
11247        [prof., dr., ir., antonius, suwanto,, m.sc]           36   
11248  [kanjeng, raden, tumengung, mas, ariya, purnam...           53   
11249  [lord, kanjeng, baginda, atep, bin, lord, dari...           76   

       avg_token_length  num_tokens  \
11245          4.833333           6   
11246          6.000000           6   
11247          5.166667           6   
11248          6.714286           7   
11249          5.416667          12   

                                         transliteration  period_freq  \
11245                 edwel yusri datuak rajo gampo alam            0   
11246          sultan amaluddin al sani perkasa alamsyah            0   
11247               prof. dr. ir. antonius suwanto, m.sc            4   
11248  kanjeng raden tumengung mas ariya purnama hadi...            0   
11249  lord kanjeng baginda atep bin lord dari segala...            0   

       dash_freq  apostrophe_freq  space_freq  
11245          0                0           5  
11246          0                0           5  
11247          0                0           5  
11248          0                0           6  
11249          0                0          11

### Malay Dataset Feature Engineering

We will implement the same features for the Malay dataset.

In [42]:
df_malay.head()

fullname          original_fullname
0              annuar rapaee              Annuar Rapaee
1                  tash yong                  Tash Yong
2                  fatmawati                  Fatmawati
3                 alto linus                 Alto Linus
4  mohamad izzat abdul halil  Mohamad Izzat Abdul Halil

In [43]:
# 1. alphabet
df_malay['alphabet'] = df_malay['fullname'].apply(get_alphabet)

# 2. unigrams, bigrams, trigrams, char_ngrams
df_malay['unigrams'] = df_malay['fullname'].apply(lambda name: list(name))
df_malay['bigrams'] = df_malay['fullname'].apply(lambda name: list(ngrams(list(name), 2)))
df_malay['trigrams'] = df_malay['fullname'].apply(lambda name: list(ngrams(list(name), 3)))
df_malay['char_ngrams'] = df_malay['unigrams'] + df_malay['bigrams'] + df_malay['trigrams']

# 3. word_ngrams
df_malay['word_ngrams'] = df_malay['fullname'].apply(lambda name: name.split())

# 4. a_hat_freq - skipped

# 5. name_length
df_malay['name_length'] = df_malay['fullname'].apply(len)

# 6. avg_token_length
tokens = df_malay['fullname'].apply(lambda name: name.split(' '))
token_lengths = tokens.apply(lambda token_list: [len(token) for token in token_list])
df_malay['avg_token_length'] = token_lengths.apply(np.mean)

# 7. num_tokens
df_malay['num_tokens'] = df_malay['fullname'].apply(lambda name: len(name.split(' ')))

# 8. transliteration
df_malay['transliteration'] = df_malay['fullname'].apply(lambda name: unidecode(name))
print("Names with special characters:")
print(pd.DataFrame(df_malay[df_malay['fullname'] != df_malay['transliteration']][['fullname', 'transliteration']]))

# 9. period_freq, dash_freq, apostrophe_freq
df_malay['period_freq'] = df_malay['fullname'].apply(lambda name: name.count('.'))
df_malay['dash_freq'] = df_malay['fullname'].apply(lambda name: name.count('-'))
df_malay['apostrophe_freq'] = df_malay['fullname'].apply(lambda name: name.count('\''))

# 10. space_freq
df_malay['space_freq'] = df_malay['fullname'].apply(lambda name: name.count(' '))

df_malay.head()

Names with special characters:
                                               fullname  \
1070                    tunku imran ibni tuanku ja’afar   
1992  sultan ahmad shah al-musta’in billah ibni alma...   
2134   tuanku ja’afar ibni almarhum tuanku abdul rahman   
2259                                     junior eldstål   
2728                   sultan abdul halim mu’adzam shah   

                                        transliteration  
1070                    tunku imran ibni tuanku ja'afar  
1992  sultan ahmad shah al-musta'in billah ibni alma...  
2134   tuanku ja'afar ibni almarhum tuanku abdul rahman  
2259                                     junior eldstal  
2728                   sultan abdul halim mu'adzam shah  


fullname          original_fullname  \
0              annuar rapaee              Annuar Rapaee   
1                  tash yong                  Tash Yong   
2                  fatmawati                  Fatmawati   
3                 alto linus                 Alto Linus   
4  mohamad izzat abdul halil  Mohamad Izzat Abdul Halil   

                                            alphabet  \
0  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...   
1  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...   
2  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   
3  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...   
4  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...   

                                            unigrams  \
0            [a, n, n, u, a, r,  , r, a, p, a, e, e]   
1                        [t, a, s, h,  , y, o, n, g]   
2                        [f, a, t, m, a, w, a, t, i]   
3                     [a, l, t, o,  , l, i, n, u, s]   
4  [m, o, h, a, m, a, d,  , i, z, z, a, t,  , a, ...   

                                             bigrams  \
0  [(a, n), (n, n), (n, u), (u, a), (a, r), (r,  ...   
1  [(t, a), (a, s), (s, h), (h,  ), ( , y), (y, o...   
2  [(f, a), (a, t), (t, m), (m, a), (a, w), (w, a...   
3  [(a, l), (l, t), (t, o), (o,  ), ( , l), (l, i...   
4  [(m, o), (o, h), (h, a), (a, m), (m, a), (a, d...   

                                            trigrams  \
0  [(a, n, n), (n, n, u), (n, u, a), (u, a, r), (...   
1  [(t, a, s), (a, s, h), (s, h,  ), (h,  , y), (...   
2  [(f, a, t), (a, t, m), (t, m, a), (m, a, w), (...   
3  [(a, l, t), (l, t, o), (t, o,  ), (o,  , l), (...   
4  [(m, o, h), (o, h, a), (h, a, m), (a, m, a), (...   

                                         char_ngrams  \
0  [a, n, n, u, a, r,  , r, a, p, a, e, e, (a, n)...   
1  [t, a, s, h,  , y, o, n, g, (t, a), (a, s), (s...   
2  [f, a, t, m, a, w, a, t, i, (f, a), (a, t), (t...   
3  [a, l, t, o,  , l, i, n, u, s, (a, l), (l, t),...   
4  [m, o, h, a, m, a, d,  , i, z, z, a, t,  , a, ...   

                      word_ngrams  name_length  avg_token_length  num_tokens  \
0                [annuar, rapaee]           13               6.0           2   
1                    [tash, yong]            9               4.0           2   
2                     [fatmawati]            9               9.0           1   
3                   [alto, linus]           10               4.5           2   
4  [mohamad, izzat, abdul, halil]           25               5.5           4   

             transliteration  period_freq  dash_freq  apostrophe_freq  \
0              annuar rapaee            0          0                0   
1                  tash yong            0          0                0   
2                  fatmawati            0          0                0   
3                 alto linus            0          0                0   
4  mohamad izzat abdul halil            0          0                0   

   space_freq  
0           1  
1           1  
2           0  
3           1  
4           3

### Indonesian Dataset: `ngram` Frequency Distributions

In [44]:
# (there may be a more efficient way of doing this)
# function that creates and returns hashmap with frequencies of all characters
# saving all possible characters as keys will help us create all possible ngrams for a given language
# this is basically unigrams frequency
def create_character_frequency_hashmap(df, names_col):
    char_freqs = {}
    for name in df[names_col]:
        for char in name:
            if char not in char_freqs.keys():
                char_freqs[char] = 1
            else:
                char_freqs[char] += 1
    return char_freqs

In [45]:
indo_char_freqs = create_character_frequency_hashmap(df_indo, 'fullname')
indo_char_freqs = dict(sorted(indo_char_freqs.items()))
indo_char_freqs

{' ': 13540,
 '"': 4,
 "'": 58,
 ',': 2,
 '-': 63,
 '.': 522,
 '/': 1,
 '2': 1,
 '9': 1,
 'a': 27270,
 'b': 2718,
 'c': 1513,
 'd': 7221,
 'e': 6867,
 'f': 1710,
 'g': 2926,
 'h': 5605,
 'i': 14578,
 'j': 2139,
 'k': 3129,
 'l': 5069,
 'm': 6743,
 'n': 12190,
 'o': 7446,
 'p': 2024,
 'q': 119,
 'r': 10548,
 's': 8250,
 't': 6285,
 'u': 7032,
 'v': 748,
 'w': 2185,
 'x': 97,
 'y': 3994,
 'z': 961,
 'á': 2,
 'ã': 3,
 'ä': 1,
 'ç': 1,
 'é': 1,
 'ë': 1,
 'í': 1,
 'ñ': 3,
 'ö': 1,
 'ú': 1,
 'ü': 1,
 'ś': 1,
 '’': 5}

In [46]:
# Using FreqDist
df_indo['unigrams_freq_dist'] = df_indo['unigrams'].apply(FreqDist)
df_indo['unigrams_freq_dist'][0]

FreqDist({'i': 2, 's': 1, 'u': 1, 'p': 1, 'r': 1, 'y': 1, 'a': 1, 'd': 1})

In [51]:
'''
Function that creates and returns all possible bigrams.
all_possible_chars: a list of all possible characters
'''
def create_all_possible_bigrams(all_possible_chars):
    all_possible_bigrams = []
    for first_char in all_possible_chars:  # first character of the current bigram
        for second_char in all_possible_chars:  # second character of the current bigram
            all_possible_bigrams.append((first_char, second_char))
    return all_possible_bigrams
    

In [61]:
all_indo_bigrams = create_all_possible_bigrams(indo_char_freqs.keys())
print(all_indo_bigrams[:20])
all_indo_bigrams[-100:]

[(' ', ' '), (' ', '"'), (' ', "'"), (' ', ','), (' ', '-'), (' ', '.'), (' ', '/'), (' ', '2'), (' ', '9'), (' ', 'a'), (' ', 'b'), (' ', 'c'), (' ', 'd'), (' ', 'e'), (' ', 'f'), (' ', 'g'), (' ', 'h'), (' ', 'i'), (' ', 'j'), (' ', 'k')]


[('ü', 'ú'),
 ('ü', 'ü'),
 ('ü', 'ś'),
 ('ü', '’'),
 ('ś', ' '),
 ('ś', '"'),
 ('ś', "'"),
 ('ś', ','),
 ('ś', '-'),
 ('ś', '.'),
 ('ś', '/'),
 ('ś', '2'),
 ('ś', '9'),
 ('ś', 'a'),
 ('ś', 'b'),
 ('ś', 'c'),
 ('ś', 'd'),
 ('ś', 'e'),
 ('ś', 'f'),
 ('ś', 'g'),
 ('ś', 'h'),
 ('ś', 'i'),
 ('ś', 'j'),
 ('ś', 'k'),
 ('ś', 'l'),
 ('ś', 'm'),
 ('ś', 'n'),
 ('ś', 'o'),
 ('ś', 'p'),
 ('ś', 'q'),
 ('ś', 'r'),
 ('ś', 's'),
 ('ś', 't'),
 ('ś', 'u'),
 ('ś', 'v'),
 ('ś', 'w'),
 ('ś', 'x'),
 ('ś', 'y'),
 ('ś', 'z'),
 ('ś', 'á'),
 ('ś', 'ã'),
 ('ś', 'ä'),
 ('ś', 'ç'),
 ('ś', 'é'),
 ('ś', 'ë'),
 ('ś', 'í'),
 ('ś', 'ñ'),
 ('ś', 'ö'),
 ('ś', 'ú'),
 ('ś', 'ü'),
 ('ś', 'ś'),
 ('ś', '’'),
 ('’', ' '),
 ('’', '"'),
 ('’', "'"),
 ('’', ','),
 ('’', '-'),
 ('’', '.'),
 ('’', '/'),
 ('’', '2'),
 ('’', '9'),
 ('’', 'a'),
 ('’', 'b'),
 ('’', 'c'),
 ('’', 'd'),
 ('’', 'e'),
 ('’', 'f'),
 ('’', 'g'),
 ('’', 'h'),
 ('’', 'i'),
 ('’', 'j'),
 ('’', 'k'),
 ('’', 'l'),
 ('’', 'm'),
 ('’', 'n'),
 ('’', 'o'),
 ('’', 'p'),